packages/flutter/lib/src/painting/cupertino_rounded_rectangle_border.dart

```diff
@@ -0,0 +1,504 @@
+
+  /// In an attempt to keep the shape of the rectangle the same regardless of its
+  /// dimension, the radius will automatically be lessened to maximize the
+  /// roundness of the resulting rectangle if its width or height is less than
+  /// 3x the radius.
+  ///
+  /// This option is best used in scenarios where a rectangle will be animated
+  /// between various different dimensions.
+  dynamicRadius,
+}
+
+/// Creates a Cupertino styled rounded rectangle - a shape similar to a rounded
+/// rectangle, but with a smoother transition from the sides to the rounded
+/// corners.
+///
+/// In this shape, the curvature of each corner over the arc is approximately
+/// a gaussian curve instead of a step function as with a traditional quarter
+/// circle round.
+///
+/// {@tool sample}
+/// ```dart
+/// Widget build(BuildContext context) {
+///   return Material(
+///     shape: CupertinoRoundedRectangleBorder(
+///       borderRadius: 28.0,
+///     ),
+///   );
+/// }
+/// ```
+/// {@end-tool}
+///
+/// See also:
+///
+/// * [RoundedRectangleBorder] Which creates a rectangle whose corners are
+///   precisely quarter circles.
+class CupertinoRoundedRectangleBorder extends ShapeBorder {
+  /// Creates a Cupertino Rounded Rectangle Border.
+  ///
+  /// The [side], [mode] and [borderRadius] arguments must not be null.
+  const CupertinoRoundedRectangleBorder({
+    this.side = BorderSide.none,
+    this.borderRadius = 0.0,
+    this.mode = RoundedRectCornerMode.dynamicShape,
+  }) : assert(side != null),
+       assert(mode != null);
+
+  /// The radius for each corner.
+  ///
+  /// The radius must be greater than or equal to 1.0.
+  final double borderRadius;
+
+  /// The style of this border.
+  final BorderSide side;
+
+  /// The corner mode of the rectangle.
+  ///
+  /// Whether or not the shape or radius will by dynamic in the event that the
+  /// width or height is smaller than 3x the radius.
+  final RoundedRectCornerMode mode;
+
+  Path _getPath(RRect rrect) {
+    double limitedRadius;
+    final double width = rrect.width;
+    final double height = rrect.height;
+    final double centerX = rrect.center.dx;
+    final double centerY = rrect.center.dy;
+    final double originX = centerX - width / 2;
+    final double originY = centerY - height / 2;
+    final double radius = math.max(1, borderRadius);
+
+    double leftX(double x) { return centerX + x * limitedRadius - width / 2; }
+    double rightX(double x) { return centerX - x * limitedRadius + width / 2; }
+    double topY(double y) { return centerY + y * limitedRadius - height / 2; }
+    double bottomY(double y) { return centerY - y * limitedRadius + height / 2; }
+    double topMidY(double y) { return originY + y * width; }
+    double bottomMidY(double y) { return originY + height - y * limitedRadius; }
+    double leftMidX(double x) { return originX + x * height; }
+    double rightMidX(double x) { return originX + width - x * limitedRadius; }
+
+    Path roundedRect1 () {
```

Comment:
Describe these algorithms a bit since it's not clear why they need to be so complicated.